pkg/driver/aws-ebs/aws_ebs_tags_controller.go

```diff
@@ -0,0 +1,441 @@
+package aws_ebs
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"sort"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorapi "github.com/openshift/api/operator/v1"
+	"github.com/openshift/csi-operator/pkg/clients"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+const (
+	awsEBSSecretNamespace = "openshift-cluster-csi-drivers"
+	awsEBSSecretName      = "ebs-cloud-credentials"
+	driverName            = "ebs.csi.aws.com"
+	tagHashAnnotationKey  = "ebs.openshift.io/volume-tags-hash"
+	batchSize             = 50
+
+	operationDelay         = 2 * time.Second
+	operationBackoffFactor = 1.2
+	operationRetryCount    = 5
+)
+
+type EBSVolumeTagsController struct {
+	name          string
+	commonClient  *clients.Clients
+	eventRecorder events.Recorder
+	failedQueue   workqueue.RateLimitingInterface
+}
+
+func NewEBSVolumeTagsController(
+	ctx context.Context,
+	name string,
+	commonClient *clients.Clients,
+	eventRecorder events.Recorder) factory.Controller {
+
+	c := &EBSVolumeTagsController{
+		name:          name,
+		commonClient:  commonClient,
+		eventRecorder: eventRecorder,
+		failedQueue:   workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 1000*time.Hour)),
+	}
+	return factory.New().WithSync(
+		c.Sync,
+	).WithInformers(
+		c.commonClient.ConfigInformers.Config().V1().Infrastructures().Informer(),
+	).WithPostStartHooks(
+		c.startFailedQueueWorker,
+	).ToController(
+		name,
+		eventRecorder,
+	)
+}
+
+func (c *EBSVolumeTagsController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.Infof("EBSVolumeTagsController sync started")
+	defer klog.Infof("EBSVolumeTagsController sync finished")
+
+	opSpec, _, _, err := c.commonClient.OperatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	infra, err := c.getInfrastructure(ctx)
+	if err != nil {
+		return err
+	}
+	var infraRegion = ""
+	if infra.Status.PlatformStatus != nil && infra.Status.PlatformStatus.AWS != nil {
+		infraRegion = infra.Status.PlatformStatus.AWS.Region
+	}
+	ec2Client, err := c.getEC2Client(ctx, infraRegion)
```

Comment:
IMO, we shouldn't create `ec2Client` here until we know we are going to need to tag PV objects. Since we are planning to backport this feature, for most infra objects, there may not even be *any* tags that needded to be synced on AWS. So, this controller will be a NO-OP for many clusters.

This means at very minimum, I would check if `infra` object has non-nil `infra.Status.PlatformStatus.AWS.ResourceTags`, before going ahead with ec2 client creation.